### ISSUE TITLE
[PSyIR] Should a Directive node have a Schedule?

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#507]
While working on #303 and thinking about how to avoid putting indices on nodes in a schedule where it did not make sense, I realised that we almost always have executable 'statements' as children of a `Schedule`. I was therefore only going to print the index of a node if its parent was a `Schedule`. However, when I tried this I realised that we do have statements as direct children of Directives. Should we therefore re-structure slightly such that Directives also contain a `Schedule`?

**Comment by hiker** [comment]
Imho if the schedule does not give us additional features, I would agree with @sergisiso and not use them - it seems to make everything more complicated.

The one potential advantage when using C could be that a schedule could always be wrapped in {}. Then #pragma omp would not have to add the {} :)  But it doesn't have any advantage for Fortran.

Re my work on directives: https://github.com/stfc/PSyclone/tree/psyir-support-directives
It is kind working, I just need more test cases from different APIs, and use omp do and parallel do as well.

**Comment by sergisiso** [comment]
@arporter The indexing just for children of Schedules is also what I was thinking because they represent a sequence which implies an order. So conceptually only Schedules have a resason to have a numbered index.

As @hiker says, in C a pragma that wraps a `{ }` block, this block should be an Schedule, but also in Fortran, the code below would be a directive for a Schedule:
```
$omp directive
 [multiple statements]
$omp end directive
```
So, I would not make Directives have Schedules in the general case in order to support and differentiate clearly the cases above.

My issue with directives is that I don't think they fit too well with the parent-child relationship, they are maybe more like annotations into a node (which can happen to be an Schedule or a single statement). The parent-child relationship might still the best way to represent it in PSyIR to work coherently with the rest of the code. But we can re-evaluate this if we need to make lots of special cases just for the Directives. Would that make transformations harder or easier to implement?

**Comment by arporter** [comment]
I don't think whether directives are a part of a language is too significant here since we're trying to make the PSyIR language agnostic. I agree that it feels wrong to have Schedules for Directives that must always have a single child (e.g. OMP Loop). However, for those that define a region (e.g. OMP Parallel) it feels natural for there to be a Schedule. Strictly speaking, there's no need for a Loop body to contain a Schedule but we decided to do that - I think to distinguish executable statements from those child nodes that hold e.g. the loop bounds? I suspect that some directives are sufficiently complex (can contain expressions, e.g. OpenACC `kernels` can have an `if(xxx)` clause) that we may need to do the same in the future.

**Comment by hiker** [comment]
I would need to study the Schedule class in a bit more details to comment more on that.

A I think somewhat related issue: I noticed one other surprising thing when trying to get a certain Assignment node:
```
# omp_sched is a Schedule, it's child is an OMPTransSomethingOrOther node etc.
omp_sched[0].children[0].loop_body[0].loop_body[0].get_kernel_schedule()
```
And that is, in some places I can just write `[x]`, in others I have to use .`children[x]`. And while I understand the distinction (`__getitem__` is defined in `Schedule`, not in `Node`), I am not sure that this makes sense from a user's point of view. Wouldn't it be better to move the __getitem__ into `Node`, so any child can be assessed using `[x]`?
This appears related to the issue at hand, since if the Directive node had a schedule, it might be different (... probably depends if it has a schedule or is a schedule ...??)

Am happy to move this as part of an upcoming pull request if you are interested.

**Comment by arporter** [comment]
After discussing #505, @rupertford and I think that, for the sake of consistency throughout the tree, everything that can have child statements should have a `Schedule`. We acknowledge that for some Nodes (e.g. OMP DO directives) this Schedule will only ever contain one child but we feel that this is a price worth paying, especially given that such directives will/can have other children describing their attributes. (e.g. some OpenACC directives can have conditional clauses.)
We can alleviate some of the resulting awkwardness by providing getter methods for the child Schedule of a node. So far we have `Loop.loop_body`, `IfBlock.if_body` and `IfBlock.else_body` that perform this function. I'm in favour of changing this slightly so that all relevant nodes have a `body` property that returns their child Schedule. An IfBlock would of course keep its additional `else_body` property.

**Comment by arporter** [comment]
After considerable discussion at the telco on 02/10/2019 we agreed to keep the existing getter methods and add a `dir_body`  getter to the `Directive` class. This will return the `Schedule` associated with the `Directive`.

**Comment by arporter** [comment]
I've implemented giving Directives a Schedule in #505. The alterations to ProfileNode and ExtractNode will be done in #542.
